src/GafferScene/Instancer.cpp

```diff
@@ -857,10 +1497,127 @@ void Instancer::compute( Gaffer::ValuePlug *output, const Gaffer::Context *conte
 		static_cast<AtomicCompoundDataPlug *>( output )->setValue( result );
 		return;
 	}
+	else if( output == variationsPlug() )
+	{
+		// Compute the number of variations by accumulating massive lists of unique hashes from all EngineDatas
+		// and then counting the total number of uniques
+
+		tbb::spin_mutex locationMutex;
+		std::vector< std::shared_ptr< EngineData::PrototypeHashes > > perLocationHashes;
+		std::shared_ptr< std::vector< InternedString > > contextVariableNames;
+		struct HashSetAccum
+		{
+			HashSetAccum(
+				const ObjectPlug *enginePlug,
+				tbb::spin_mutex &locationMutex,
+				std::vector< std::shared_ptr< EngineData::PrototypeHashes > > &perLocationHashes,
+				std::shared_ptr< std::vector< InternedString > > &contextVariableNames
+			):
+				m_enginePlug( enginePlug ), m_locationMutex( locationMutex ),
+				m_perLocationHashes( perLocationHashes ), m_contextVariableNames( contextVariableNames )
+			{
+			}
+			bool operator()( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+			{
+				ConstEngineDataPtr engine = boost::static_pointer_cast<const EngineData>( m_enginePlug->getValue() );
+				std::shared_ptr< EngineData::PrototypeHashes > locationHashes = engine->uniquePrototypeHashes();
+
+				tbb::spin_mutex::scoped_lock lock( m_locationMutex );
+				m_perLocationHashes.push_back( locationHashes );
+				if( !m_contextVariableNames )
+				{
+					// The context variable names are guaranteed to match for all EngineData, so we just
+					// grab them from the first EngineData we process
```

Comment:
Isn't `scene:path` in the context when we compute `EngineData`, so couldn't a user enable/disable context variables on a per-location basis? And are you convinced that using a vector of values and a vector of names is actually important for performance here? For this naive reader at least, a map would be far easier to understand, and it seems plausible that indexing overhead really might not be something to worry about in this case.

Reply:
Switched.